nomad/state/state_store.go

```diff
@@ -3986,6 +4015,29 @@ func (s *StateStore) schedulerSetConfigTxn(idx uint64, tx *memdb.Txn, config *st
 	return nil
 }
 
+func (s *StateStore) setClusterMetadata(txn *memdb.Txn, meta *structs.ClusterMetadata) error {
+	// Check for an existing config, if it exists, sanity check the ClusterID matches
+	existing, err := txn.First("cluster_meta", "id")
+	if err != nil {
+		return fmt.Errorf("failed cluster meta lookup: %v", err)
+	}
+
+	if existing != nil {
+		existingClusterID := existing.(*structs.ClusterMetadata).ClusterID
+		if meta.ClusterID != existingClusterID {
+			// there is a bug in clusterID detection
+			return fmt.Errorf("refusing to set new cluster id, previous: %s, new: %s", existingClusterID, meta.ClusterID)
```

Comment:
Is this what Consul does? This error gets logged and operation continues normally. This seems like an opportunity for a node to detect a misconfiguration (eg server from staging cluster told to join prod cluster) and refuse further operations (perhaps by panicking or shutting down gracefully?!).

If there is an unambiguous strong action we can take here, let's file a follow-up ticket and worry about it later. It's an enhancement on top of this functionality.